### ISSUE TITLE
Prohibit optimizer.zero_grad when use two different DistributedOptimizer

### ISSUE THREAD

**Issue by ldd91** [horovod/horovod#1417]
**Environment:**
1. Framework: (TensorFlow, Keras, PyTorch, MXNet)
2. Framework version:
3. Horovod version:0.18.1
4. MPI version:3.0.4
5. CUDA version:10.1
6. NCCL version:
7. Python version:python3
8. OS and version:linux
9. GCC version:

**Checklist:**
1. Did you search issues to find if somebody asked this question before?
2. If your question is about hang, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/running.rst)?
3. If your question is about docker, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/docker.rst)?
4. Did you check if you question is answered in the [troubleshooting guide](https://github.com/horovod/horovod/blob/master/docs/troubleshooting.rst)?

**Your question:**
Please ask your question here.
@gkanwar Hi,I use two optimizer 
w_optimiter =  torch.optim.SGD(param=param_except_thetas,...)
theta_optimizer = torch.optim.Adam(param=theta_params,...)
horovod_w_optimiter = hvd.DistributedOptimizer(w_optimizer,...)
horovod_theta_optimizer = hvd.DistributedOptimizer(theta_optimizer,..)
...
self._training_step(model, train_w_loader, self.w_optimizer, epoch, info_for_logging="_w_step_")
self.w_scheduler.step()
self.w_optimizer.synchronize()
self._training_step(model, train_theta_loader, self.theta_optimizer, epoch,info_for_logging="_w_step_")
 however an error has been encountered,whth the log shows:
FIle ... line 199, in zero_grad
raise AssertionError(" optimizer.zero_grad() was called after loss.backward()"
AssertionError(" optimizer.zero_grad() was called after loss.backward() but before optimizer.step()

**Comment by tgaddair** [comment]
Hey @ldd91, apologies for the late response.  Can you create a gist with a more complete repro for more context?

**Comment by ldd91** [comment]
@tgaddair  thanks for your replay,I see in your example all have only one optimiter,but in my project we use two optimiter,it is there any different with the one optimiter

**Comment by ldd91** [comment]
@tgaddair this is the origin code that use only one gpu and no horovod：
import numpy as np
import torch
from torch import nn
from tensorboardX import SummaryWriter
from scipy.special import softmax
import argparse

from general_functions.dataloaders import get_loaders, get_test_loader
from general_functions.utils import get_logger, weights_init, load, create_directories_from_list, \
                                    check_tensor_in_list, writh_new_ARCH_to_fbnet_modeldef
from supernet_functions.lookup_table_builder import LookUpTable
from supernet_functions.model_supernet import FBNet_Stochastic_SuperNet, SupernetLoss
from supernet_functions.training_functions_supernet import TrainerSupernet
from supernet_functions.config_for_supernet import CONFIG_SUPERNET
from fbnet_building_blocks.fbnet_modeldef import MODEL_ARCH
    
parser = argparse.ArgumentParser("action")
parser.add_argument('--train_or_sample', type=str, default='', \
                    help='train means training of the SuperNet, sample means sample from SuperNet\'s results')
parser.add_argument('--architecture_name', type=str, default='', \
                    help='Name of an architecture to be sampled')
parser.add_argument('--hardsampling_bool_value', type=str, default='True', \
                    help='If not False or 0 -> do hardsampling, else - softmax sampling')
args = parser.parse_args()

def train_supernet():
    manual_seed = 1
    np.random.seed(manual_seed)
    torch.manual_seed(manual_seed)
    torch.cuda.manual_seed_all(manual_seed)
    torch.backends.cudnn.benchmark = True

    create_directories_from_list([CONFIG_SUPERNET['logging']['path_to_tensorboard_logs']])
    
    logger = get_logger(CONFIG_SUPERNET['logging']['path_to_log_file'])
    writer = SummaryWriter(log_dir=CONFIG_SUPERNET['logging']['path_to_tensorboard_logs'])
    
    #### LookUp table consists all information about layers
    lookup_table = LookUpTable(calulate_latency=CONFIG_SUPERNET['lookup_table']['create_from_scratch'])
    
    #### DataLoading
    train_w_loader, train_thetas_loader = get_loaders(CONFIG_SUPERNET['dataloading']['w_share_in_train'],
                                                      CONFIG_SUPERNET['dataloading']['batch_size'],
                                                      CONFIG_SUPERNET['dataloading']['path_to_save_data'],
                                                      logger)
    test_loader = get_test_loader(CONFIG_SUPERNET['dataloading']['batch_size'],
                                  CONFIG_SUPERNET['dataloading']['path_to_save_data'])
    
    #### Model
    model = FBNet_Stochastic_SuperNet(lookup_table, cnt_classes=10).cuda()
    model = model.apply(weights_init)
    model = nn.DataParallel(model, device_ids=[0])
    
    #### Loss, Optimizer and Scheduler
    criterion = SupernetLoss().cuda()

    thetas_params = [param for name, param in model.named_parameters() if 'thetas' in name]
    params_except_thetas = [param for param in model.parameters() if not check_tensor_in_list(param, thetas_params)]

    w_optimizer = torch.optim.SGD(params=params_except_thetas,
                                  lr=CONFIG_SUPERNET['optimizer']['w_lr'], 
                                  momentum=CONFIG_SUPERNET['optimizer']['w_momentum'],
                                  weight_decay=CONFIG_SUPERNET['optimizer']['w_weight_decay'])
    
    theta_optimizer = torch.optim.Adam(params=thetas_params,
                                       lr=CONFIG_SUPERNET['optimizer']['thetas_lr'],
                                       weight_decay=CONFIG_SUPERNET['optimizer']['thetas_weight_decay'])

    last_epoch = -1
    w_scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(w_optimizer,
                                                             T_max=CONFIG_SUPERNET['train_settings']['cnt_epochs'],
                                                             last_epoch=last_epoch)
    
    #### Training Loop
    trainer = TrainerSupernet(criterion, w_optimizer, theta_optimizer, w_scheduler, logger, writer)
    trainer.train_loop(train_w_loader, train_thetas_loader, test_loader, model)

if __name__ == "__main__":
    assert args.train_or_sample in ['train', 'sample']
    if args.train_or_sample == 'train':
        train_supernet()
    elif args.train_or_sample == 'sample':
        assert args.architecture_name != '' and args.architecture_name not in MODEL_ARCH
        hardsampling = False if args.hardsampling_bool_value in ['False', '0'] else True
        sample_architecture_from_the_supernet(unique_name_of_arch=args.architecture_name, hardsampling=hardsampling)

**Comment by RayOct18** [comment]
Have anyone solve this problem?
I have same situation.
Works fine with the optimizer, but two optimizers cause this problem.

**Comment by howardgriffin** [comment]
I have the same problem when I train a GAN. Could anyone help me sovle it?

**Comment by tgaddair** [comment]
Hey @ldd91 @RayOct18 @howardgriffin, the issue here seems to be that every time `loss.backwards()` is called, all optimizers receive gradients, and if `optimizer.step()` or `optimizer.synchronize()` isn't called before `optimizer.zero_grad()` after that, then this exception is raised.

You can workaround this by calling `optimizer.synchronize()` before calling `optimizer.zero_grad()`.  This will cause the optimizer to discard the gradients for this step (which you want to ignore).

Here's a minimum example that reproed the issue.  Once I added the `theta_optimizer.synchronize()` call below, the exception went away. Let me know if this is consistent with what everyone here is trying to accomplish:

```
import torch
import torch.nn as nn
import torch.nn.functional as F
import horovod.torch as hvd

class Model(nn.Module):
    def __init__(self, a, b):
        super(Model, self).__init__()
        self.a = nn.Parameter(a)
        self.b = nn.Parameter(b)

    def forward(self, x):
        return self.a * x + self.b

hvd.init()

a = torch.rand(2)
b = torch.rand(2)

model = Model(a, b)

w_params = [(name, param) for name, param in model.named_parameters() if 'a' in name]
theta_params = [(name, param) for name, param in model.named_parameters() if 'b' in name]

w_optimizer = torch.optim.SGD([p for n, p in w_params], lr=0.01)
w_optimizer = hvd.DistributedOptimizer(w_optimizer, named_parameters=w_params)

theta_optimizer = torch.optim.Adam([p for n, p in theta_params], lr=0.01)
theta_optimizer = hvd.DistributedOptimizer(theta_optimizer, named_parameters=theta_params)

N, D_in, H, D_out = 64, 2, 10, 2
x = torch.randn(N, D_in).requires_grad_()
y = torch.randn(N, D_out).requires_grad_()

y_pred = model(x)
loss = F.mse_loss(y_pred, y, size_average=False)
w_optimizer.zero_grad()
loss.backward()
w_optimizer.step()

# Note: theta_optimizer
theta_optimizer.synchronize()

y_pred = model(x)
loss = F.mse_loss(y_pred, y, size_average=False)
theta_optimizer.zero_grad()
loss.backward()
theta_optimizer.step()
```

**Comment by ldd91** [comment]
Thank you for your reply,i will have a try,and show the result.

**Comment by ParkGeonYeong** [comment]
@tgaddair I got no error but a warning: 
optimizer.step() called without optimizer.skip_synchronize() context after optimizer.synchronize(). It says it may cause training slowdown. Above your example, is it okay just do a normal optimizer.step() or add skip_synchronize()?

**Comment by cfanyyx** [comment]
Same warning as ParkGeonYeong. Could anyone give a solution?

**Comment by cheneyoung** [comment]
> @tgaddair I got no error but a warning:
> optimizer.step() called without optimizer.skip_synchronize() context after optimizer.synchronize(). It says it may cause training slowdown. Above your example, is it okay just do a normal optimizer.step() or add skip_synchronize()?

I meet the same problem when I training on styleGAN, anyone could give a solution?

**Comment by ParkGeonYeong** [comment]
@cfanyyx, @cheneyoung  In my case, i don't see any qualitative problems without using .skip_synchronize(). It seems fine.

**Comment by cheneyoung** [comment]
> @cfanyyx, @cheneyoung In my case, i don't see any qualitative problems without using .skip_synchronize(). It seems fine.

How did you use .skip_synchronize()? Below is my code:
`# update generator weights
        optimizer_G.synchronize()
        optimizer_G.zero_grad()
        if opt.fp16:                                
            with amp.scale_loss(loss_G, optimizer_G) as scaled_loss: scaled_loss.backward()                
        else:
            loss_G.backward()          
        optimizer_G.step()
        optimizer_G.skip_synchronize()

        # update discriminator weights
        optimizer_D.synchronize()
        optimizer_D.zero_grad()
        if opt.fp16:                                
            with amp.scale_loss(loss_D, optimizer_D) as scaled_loss: scaled_loss.backward()                
        else:
            loss_D.backward()        
        optimizer_D.step()  
        optimizer_D.skip_synchronize() `
